platform/o.n.agent/src/org/netbeans/agent/TransferHandlerActionsNonFinalTransformer.java

```diff
@@ -0,0 +1,107 @@
+import org.objectweb.asm.Opcodes;
+
+/**
+ * Transform javax.swing.TransferHandler to enable replacing the copy, cut and
+ * paste actions and replace them with their NetBeans implementations.
+ *
+ * The transformer can be debugged:
+ *
+ * - The system property "org.netbeans.agent.TransferHandlerActionsNonFinalTransformer.disable"
+ *   takes a boolean to disable the transformer. I.e. running NetBeans as
+ *
+ *   NBBASEPATH/bin/netbeans64.exe -J-Dorg.netbeans.agent.TransferHandlerActionsNonFinalTransformer.disable=true
+ *
+ *   will run without this.
+ *
+ * - The system property "org.netbeans.agent.TransferHandlerActionsNonFinalTransformer.dumpPath"
+ *   takes a string. The string is expected to be a path and the transformed
+ *   class data will be dumped at that location. It is expected, that a path
+ *   including the filename is specified.
+ */
+public class TransferHandlerActionsNonFinalTransformer implements ClassFileTransformer {
+
+    private static final String DEBUG_DUMP_TRANSFORMED_CLASS = System.getProperty("org.netbeans.agent.TransferHandlerActionsNonFinalTransformer.dumpPath");
+    private static final boolean DEBUG_DISABLE_TRANSFORMER = Boolean.getBoolean("org.netbeans.agent.TransferHandlerActionsNonFinalTransformer.disable");
+
+    private static Logger getLog() {
+        return Logger.getLogger(TransferHandlerActionsNonFinalTransformer.class.getName());
+    }
+
+    @Override
+    public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+        if (DEBUG_DISABLE_TRANSFORMER) {
+            return null;
+        }
+        // We assume, that we are loaded early enough before AWT is initialzed,
+        // so we don't expect to need to retransform
+        if (classBeingRedefined != null) {
+            getLog().log(Level.WARNING, "Attempted retransform for: {0} (Ignored)", new Object[]{className});
+        }
+        // Only transform TransferHandler class - ignore all other classes
+        if ("javax/swing/TransferHandler".equals(className)) {
+            getLog().log(Level.INFO, "Transforming: {0}", new Object[]{className});
+            try {
+                ClassReader cr = new ClassReader(classfileBuffer);
+                ClassWriter cw = new ClassWriter(cr, ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS);
+                ClassVisitor cv = new ClassVisitor(Opcodes.ASM9, cw) {
+                    @Override
+                    public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
+                        switch(name) {
+                            case "cutAction":
+                            case "copyAction":
+                            case "pasteAction":
+                                return super.visitField(access & (~ Opcodes.ACC_FINAL), name, descriptor, signature, value);
```

Comment:
Sorry, I don't like this. Removing the `final` modifier will at least make sure that deep reflection will work on the field, but it is still adding (new) deep reflection against JDK class, which not very good. And removing the `final` modifier also feels like a highly suspicious thing to me.

Given we have a transforming agent, is there a reason why not change `TransferAction.getClipboard()` to do whatever is needed? Or something else, that would be less problematic than using deep reflection on a field from the JDK, and removing the `final` modifier to make the reflection work.

Reply:
Removed the Transformer, so does not apply anymore.